### ISSUE TITLE
Stop metric made increasing memory usage

### ISSUE THREAD

**Issue by GodBlessZhk** [dmlc/mxnet#6892]
## Environment info
Ubuntu 16.04, GCC-5.4.0, MXNet 0.9.5, Python 2.7

## Error Message
I'd like to train a model without a metric, so I use the code just as follows. But I find that the memory usage keeps increasing until it's full. 
```
eval_metrics = mx.metric.CompositeEvalMetric()
train_data = mx.io.PrefetchingIter(train_data)
mod.fit(train_data, eval_metric=eval_metrics, ...)
```
Then I define a class named 'NoMetric' as follows and use it as the metric.
```
class NoMetric(mx.metric.EvalMetric):
    def __init__(self):
        super(NoMetric, self).__init__('NoMetric')

    def update(self, labels, preds):
        self.sum_metric += 1
        self.num_inst += 1
```
But the memory still keeps increasing, so I try this:
```
class NoMetric(mx.metric.EvalMetric):
    def __init__(self):
        super(NoMetric, self).__init__('NoMetric')
        self.e2e = config.TRAIN.END2END
        self.rcnn_ohem = config.TRAIN.RCNN_OHEM
        self.pred, self.label = get_rcnn_names()

    def update(self, labels, preds):  	
    	bbox_loss = preds[self.pred.index('rcnn_bbox_loss')].asnumpy()
        if self.e2e:
            label = preds[self.pred.index('rcnn_label')].asnumpy()
        else:
            label = labels[self.label.index('rcnn_label')].asnumpy()
        
        self.sum_metric += 1
        self.num_inst += 1
```
Luckily, the memory usage doesn't increase anymore. But I want to know why I must call asnumpy()?

## Latest

I find that whether I use `train_data = mx.io.PrefetchingIter(train_data)` or not, the memory keeps increasing. Similarly, if I use multi-process to read data in a queue for acceleration, this problem will also appear.

The latest version of MXNet doesn't fix my bug.

**Comment by Zehaos** [comment]
I found that it is extremely fast when training RCNN with metric disabled.

**Comment by CNevd** [comment]
I’ve run [train_cifar10.py](https://github.com/dmlc/mxnet/blob/master/example/image-classification/train_cifar10.py) with ```eval_metrics = mx.metric.CompositeEvalMetric()``` however the memory is normal on my machine  you can try a new version of MXNet?

**Comment by GodBlessZhk** [comment]
@CNevd I've updated the error message. 
I really find that the memory is normal if I don't use PrefetchIter. But I still need to accelerate the training procedure such as I/O. I will have a try for the new version of MXNet.

**Comment by GodBlessZhk** [comment]
@CNevd I've tried the latest version of MXNet, but nothing changed. 
I'm sorry to say that I made a mistake yesterday. And now I find there's no difference whether I use PrefetchIter or not. The memory usage will keep increasing if I disable metric.

**Comment by CNevd** [comment]
Could you post your code  to reproduce it?

**Comment by GodBlessZhk** [comment]
@CNevd I have found the reason. If using metric, it will wait for the data processed by GPU and then start a new iteration. But if not using metric, it will keep loading data whether GPU finishes processing data or not. So the previous data would not be freed.
Thank you very much, I will close the issue.

**Comment by nikrao** [comment]
Hello, @GodBlessZhk @CNevd . other than the solution above, has this issue been addressed in a more principled way?What if we don't want to call metric after every iteration?
